# Data Diggers Team Assignment Analytics Applications

In [1]:
# packages
import pandas as pd
import numpy as np
import json
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# read csv files
try:
    charging_sessions_df = pd.read_csv('resources/charging_sessions.csv')
    print("Successfully imported charging_sessions.csv")
except FileNotFoundError:
    print("Error: charging_sessions.csv not found.")

try:
    weather_burbank_airport_df = pd.read_csv('resources/weather_burbank_airport.csv')
    print("Successfully imported weather_burbank_airport.csv")
except FileNotFoundError:
    print("Error: weather_burbank_airport.csv not found.")

Successfully imported charging_sessions.csv
Successfully imported weather_burbank_airport.csv


In [211]:
charging_sessions_df

Unnamed: 0                        id             connectionTime  \
0               0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1               1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2               2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3               3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4               4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...           ...                       ...                        ...   
66445       10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446       10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447       10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448       10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449       10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN  

[66450 rows x 13 columns]

## Data Collection and Preparation

In the upcoming section, we want to observe how many erroneous datasets we have and how we should handle them.

### Format
We first observe that the .csv file's first column is an unnamed column with integers. Since there is no description in the .pdf file for the team assignment, we will drop these incoherent values and get rid of the very first column.

In [212]:
formatted_df = charging_sessions_df.copy().iloc[:, 1:]
formatted_df.head(3)

id             connectionTime  \
0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...

Furthermore, if we take a look at the ***userInputs*** column, we see that the provided data is in *json*-format.
For easier handling, we want to transform this column to an appropriate format for our analysis.
We will thus create a new column for each field in the json.

In [213]:
def clean_and_parse_json(value):
    if isinstance(value, str):
        cleaned_value = value.replace("'", '"')
        cleaned_value = cleaned_value.replace("True", "true").replace("False", "false")

        if cleaned_value.startswith("[") and cleaned_value.endswith("]"):
            try:
                return json.loads(cleaned_value)
            except json.JSONDecodeError as e:
                print(f"Error parsing JSON: {e}")
                return None
        else:
            return None
    return None


formatted_df['parsedJson'] = formatted_df['userInputs'].apply(clean_and_parse_json)


all_keys = set()
for row in formatted_df['parsedJson'].dropna():
    if isinstance(row, list) and row:
        all_keys.update(row[0].keys())

for field in all_keys:
    formatted_df[field] = formatted_df['parsedJson'].apply(
        lambda x: x[0].get(field) if isinstance(x, list) and x else np.nan
    )

formatted_df = formatted_df.drop(columns=['userInputs', 'parsedJson'])

formatted_df.head(3)

id             connectionTime  \
0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   

              timezone  userID paymentRequired  minutesAvailable  \
0  America/Los_Angeles   194.0            True             463.0   
1  America/Los_Angeles  4275.0            True             595.0   
2  America/Los_Angeles   344.0            True              60.0   

                      modifiedAt             requestedDeparture  \
0  Thu, 02 Jan 2020 13:09:39 GMT  Thu, 02 Jan 2020 20:51:54 GMT   
1  Thu, 02 Jan 2020 13:37:11 GMT  Thu, 02 Jan 2020 23:31:50 GMT   
2  Thu, 02 Jan 2020 13:57:17 GMT  Thu, 02 Jan 2020 14:56:35 GMT   

   milesRequested  kWhRequested  WhPerMile  
0           100.0          25.0      250.0  
1           250.0          70.0      280.0  
2            20.0           8.0      400.0

Now we take a look at the datatypes in our dataframe:

In [214]:
print(formatted_df.dtypes)

id                     object
connectionTime         object
disconnectTime         object
doneChargingTime       object
kWhDelivered          float64
sessionID              object
siteID                  int64
spaceID                object
stationID              object
timezone               object
userID                float64
paymentRequired        object
minutesAvailable      float64
modifiedAt             object
requestedDeparture     object
milesRequested        float64
kWhRequested          float64
WhPerMile             float64
dtype: object


For easier handling during our analysis, it might be benificial to convert the values of **connectionTime**, **disconnectTime**, **doneChargingtime**, **modifiedAt** and **requestedDeparture** from type object to datetime.

In [215]:
formatted_df['connectionTime'] = pd.to_datetime(formatted_df['connectionTime'], utc=False, errors='coerce')
formatted_df['disconnectTime'] = pd.to_datetime(formatted_df['disconnectTime'], utc=False, errors='coerce')
formatted_df['doneChargingTime'] = pd.to_datetime(formatted_df['doneChargingTime'], utc=False, errors='coerce')
formatted_df['modifiedAt'] = pd.to_datetime(formatted_df['modifiedAt'], utc=False, errors='coerce')
formatted_df['requestedDeparture'] = pd.to_datetime(formatted_df['requestedDeparture'], utc=False, errors='coerce')
print(formatted_df.dtypes)

id                                 object
connectionTime        datetime64[ns, UTC]
disconnectTime        datetime64[ns, UTC]
doneChargingTime      datetime64[ns, UTC]
kWhDelivered                      float64
sessionID                          object
siteID                              int64
spaceID                            object
stationID                          object
timezone                           object
userID                            float64
paymentRequired                    object
minutesAvailable                  float64
modifiedAt                 datetime64[ns]
requestedDeparture         datetime64[ns]
milesRequested                    float64
kWhRequested                      float64
WhPerMile                         float64
dtype: object


In [216]:
formatted_df.head(3)

id            connectionTime  \
0  5e23b149f9af8b5fe4b973cf 2020-01-02 13:08:54+00:00   
1  5e23b149f9af8b5fe4b973d0 2020-01-02 13:36:50+00:00   
2  5e23b149f9af8b5fe4b973d1 2020-01-02 13:56:35+00:00   

             disconnectTime          doneChargingTime  kWhDelivered  \
0 2020-01-02 19:11:15+00:00 2020-01-02 17:31:35+00:00        25.016   
1 2020-01-02 22:38:21+00:00 2020-01-02 20:18:05+00:00        33.097   
2 2020-01-03 00:39:22+00:00 2020-01-02 16:35:06+00:00         6.521   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   

              timezone  userID paymentRequired  minutesAvailable  \
0  America/Los_Angeles   194.0            True             463.0   
1  America/Los_Angeles  4275.0            True             595.0   
2  America/Los_Angeles   344.0            True              60.0   

           modifiedAt  requestedDeparture  milesRequested  kWhRequested  \
0 2020-01-02 13:09:39 2020-01-02 20:51:54           100.0          25.0   
1 2020-01-02 13:37:11 2020-01-02 23:31:50           250.0          70.0   
2 2020-01-02 13:57:17 2020-01-02 14:56:35            20.0           8.0   

   WhPerMile  
0      250.0  
1      280.0  
2      400.0

Let's also take a look at the amount of distinct values for each column.

In [217]:
print(formatted_df.nunique())

id                    65037
connectionTime        64839
disconnectTime        64906
doneChargingTime      60637
kWhDelivered          25629
sessionID             65037
siteID                    2
spaceID                 107
stationID               107
timezone                  1
userID                 1006
paymentRequired           1
minutesAvailable        842
modifiedAt            47704
requestedDeparture    47773
milesRequested          106
kWhRequested           1201
WhPerMile               178
dtype: int64


We observe that for the columns **timezone** and **paymentRequired**, there is only **one** unique value.
Thus, we can drop them as they don't add any value to our analysis since we they can't be used to compare / differentiate rows.

In [218]:
formatted_df = formatted_df.drop(columns=['timezone','paymentRequired'])

### Duplicates
First we want to find out whether any duplicates exist which we can check via the **id** field of each dataset.
If so, those duplicates shall be removed.

In [219]:
len(formatted_df[formatted_df['id'].duplicated()])

1413

As we can see, there are **1413** duplicated rows. We will thus remove them for further analysis.

In [220]:
formatted_df = formatted_df.drop_duplicates(subset=['id'])
formatted_df

id            connectionTime  \
0      5e23b149f9af8b5fe4b973cf 2020-01-02 13:08:54+00:00   
1      5e23b149f9af8b5fe4b973d0 2020-01-02 13:36:50+00:00   
2      5e23b149f9af8b5fe4b973d1 2020-01-02 13:56:35+00:00   
3      5e23b149f9af8b5fe4b973d2 2020-01-02 13:59:58+00:00   
4      5e23b149f9af8b5fe4b973d3 2020-01-02 14:00:01+00:00   
...                         ...                       ...   
65032  5d2fbdd3f9af8b4d0dd0d54a 2019-07-01 21:49:12+00:00   
65033  5d2fbdd3f9af8b4d0dd0d54b 2019-07-01 21:58:45+00:00   
65034  5d2fbdd3f9af8b4d0dd0d54c 2019-07-01 22:02:21+00:00   
65035  5d2fbdd3f9af8b4d0dd0d54d 2019-07-01 22:23:44+00:00   
65036  5d2fbdd3f9af8b4d0dd0d54e 2019-07-01 22:36:08+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
0     2020-01-02 19:11:15+00:00 2020-01-02 17:31:35+00:00        25.016   
1     2020-01-02 22:38:21+00:00 2020-01-02 20:18:05+00:00        33.097   
2     2020-01-03 00:39:22+00:00 2020-01-02 16:35:06+00:00         6.521   
3     2020-01-02 16:38:39+00:00 2020-01-02 15:18:45+00:00         2.355   
4     2020-01-02 22:08:40+00:00 2020-01-02 18:17:30+00:00        13.375   
...                         ...                       ...           ...   
65032 2019-07-02 01:37:18+00:00 2019-07-01 23:14:33+00:00         8.399   
65033 2019-07-02 00:39:48+00:00 2019-07-02 00:40:21+00:00        16.864   
65034 2019-07-02 00:58:50+00:00 2019-07-02 00:59:23+00:00        18.335   
65035 2019-07-02 02:03:36+00:00 2019-07-02 02:04:06+00:00        22.815   
65036 2019-07-02 01:10:52+00:00 2019-07-02 00:24:32+00:00        10.241   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
65032  1_1_179_798_2019-07-01 21:49:11.873404       1  AG-3F18  1-1-179-798   
65033  1_1_179_794_2019-07-01 21:58:44.571011       1  AG-3F20  1-1-179-794   
65034  1_1_191_807_2019-07-01 22:02:20.810735       1  AG-4F47  1-1-191-807   
65035  1_1_179_781_2019-07-01 22:23:32.496137       1  AG-3F31  1-1-179-781   
65036  1_1_193_819_2019-07-01 22:35:56.104381       1  AG-1F06  1-1-193-819   

       userID  minutesAvailable          modifiedAt  requestedDeparture  \
0       194.0             463.0 2020-01-02 13:09:39 2020-01-02 20:51:54   
1      4275.0             595.0 2020-01-02 13:37:11 2020-01-02 23:31:50   
2       344.0              60.0 2020-01-02 13:57:17 2020-01-02 14:56:35   
3      1117.0              65.0 2020-01-02 14:00:03 2020-01-02 15:04:58   
4       334.0             504.0 2020-01-02 14:00:13 2020-01-02 22:24:01   
...       ...               ...                 ...                 ...   
65032  1346.0              36.0 2019-07-01 21:49:37 2019-07-01 22:25:12   
65033   364.0             234.0 2019-07-01 21:59:09 2019-07-02 01:52:45   
65034  2050.0             309.0 2019-07-01 22:02:37 2019-07-02 03:11:21   
65035  1626.0              99.0 2019-07-01 22:23:43 2019-07-02 00:02:44   
65036  2299.0             191.0 2019-07-01 22:36:08 2019-07-02 01:47:08   

       milesRequested  kWhRequested  WhPerMile  
0               100.0         25.00      250.0  
1               250.0         70.00      280.0  
2                20.0          8.00      400.0  
3                20.0          8.00      400.0  
4                40.0         16.00      400.0  
...               ...           ...        ...  
65032            80.0         32.00      400.0  
65033           100.0         40.00      400.0  
65034            90.0         29.97      333.0  
65035           170.0         34.00      200.0  
65036 

### Missing Data
We now want to check the data for missing values

In [221]:
formatted_df.isnull().sum()

id                        0
connectionTime            0
disconnectTime            0
doneChargingTime       4087
kWhDelivered              0
sessionID                 0
siteID                    0
spaceID                   0
stationID                 0
userID                17215
minutesAvailable      17215
modifiedAt            17215
requestedDeparture    17215
milesRequested        17215
kWhRequested          17215
WhPerMile             17215
dtype: int64

As we can see, there are **4087** rows with a missing **doneChargingTime** value as well as **17215** rows with missing **userID** and **userInputs** value. The latter is expected as charging sessions may also be carried out by unregistered users. Thus, we only have to take care of the **doneChargingTime** values. We try to find out whether this is related to a specific other value.

In [222]:
missing_done_charging_time_df = formatted_df[formatted_df.doneChargingTime.isnull()]
space_id_value_counts = missing_done_charging_time_df['stationID'].value_counts()
space_id_value_counts

stationID
2-39-81-4550    1125
1-1-178-823       89
1-1-193-820       77
1-1-178-828       67
1-1-193-827       63
                ... 
2-39-129-563       2
2-39-126-560       1
2-39-125-559       1
2-39-130-564       1
2-39-82-384        1
Name: count, Length: 100, dtype: int64

The stationID **2-39-81-4550** seems to be faulty when it comes to this value. When we look at its records:

In [223]:
print("stationID 2-39-81-4550")
print(f"null values: {missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']['doneChargingTime'].isnull().sum()}")
print(f"non null values: {missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']['doneChargingTime'].notnull().sum()}")

stationID 2-39-81-4550
null values: 1125
non null values: 0


In [224]:
missing_done_charging_time_df[missing_done_charging_time_df.stationID == '2-39-81-4550']

id            connectionTime  \
46086  6004e49bf9af8b546d1d6e03 2021-01-01 22:29:34+00:00   
46089  6004e49bf9af8b546d1d6e06 2021-01-02 02:03:29+00:00   
46090  6004e49bf9af8b546d1d6e07 2021-01-02 02:53:51+00:00   
46091  6004e49bf9af8b546d1d6e08 2021-01-02 04:34:46+00:00   
46092  6006361cf9af8b569d876177 2021-01-02 15:07:37+00:00   
...                         ...                       ...   
51579  6000f01bf9af8b4e2987106a 2020-12-30 06:35:39+00:00   
51580  6002419cf9af8b50514f4c3c 2020-12-30 18:17:31+00:00   
51582  6002419cf9af8b50514f4c3e 2020-12-30 22:09:41+00:00   
51584  6003931bf9af8b5261c8c793 2020-12-31 16:17:43+00:00   
51586  6003931bf9af8b5261c8c795 2020-12-31 17:42:03+00:00   

                 disconnectTime doneChargingTime  kWhDelivered  \
46086 2021-01-01 23:20:56+00:00              NaT        35.503   
46089 2021-01-02 02:19:50+00:00              NaT        11.827   
46090 2021-01-02 03:23:48+00:00              NaT         7.646   
46091 2021-01-02 05:22:23+00:00              NaT        31.138   
46092 2021-01-02 15:55:46+00:00              NaT        12.768   
...                         ...              ...           ...   
51579 2020-12-30 07:30:14+00:00              NaT        30.430   
51580 2020-12-30 19:09:46+00:00              NaT        29.623   
51582 2020-12-30 22:33:48+00:00              NaT         6.331   
51584 2020-12-31 17:09:35+00:00              NaT        24.903   
51586 2020-12-31 18:29:00+00:00              NaT        20.526   

                                     sessionID  siteID   spaceID  \
46086  2_39_81_4550_2021-01-01 22:29:30.587407       2  11900388   
46089  2_39_81_4550_2021-01-02 02:03:26.186580       2  11900388   
46090  2_39_81_4550_2021-01-02 02:53:45.713992       2  11900388   
46091  2_39_81_4550_2021-01-02 04:34:39.442790       2  11900388   
46092  2_39_81_4550_2021-01-02 15:07:37.007402       2  11900388   
...                                        ...     ...       ...   
51579  2_39_81_4550_2020-12-30 06:35:38.926587       2  11900388   
51580  2_39_81_4550_2020-12-30 18:17:30.573937       2  11900388   
51582  2_39_81_4550_2020-12-30 22:09:41.373141       2  11900388   
51584  2_39_81_4550_2020-12-31 16:17:42.790605       2  11900388   
51586  2_39_81_4550_2020-12-31 17:42:03.239398       2  11900388   

          stationID   userID  minutesAvailable          modifiedAt  \
46086  2-39-81-4550   9561.0             480.0 2021-01-01 22:29:34   
46089  2-39-81-4550   9283.0             270.0 2021-01-02 02:03:28   
46090  2-39-81-4550   9283.0             270.0 2021-01-02 02:53:51   
46091  2-39-81-4550    915.0             721.0 2021-01-02 04:34:46   
46092  2-39-81-4550   7132.0             480.0 2021-01-02 15:08:06   
...             ...      ...               ...                 ...   
51579  2-39-81-4550   1424.0             663.0 2020-12-30 06:35:54   
51580  2-39-81-4550    268.0             480.0 2020-12-30 18:17:51   
51582  2-39-81-4550  10090.0             480.0 2020-12-30 22:09:58   
51584  2-39-81-4550   1039.0             288.0 2020-12-31 16:18:25   
51586  2-39-81-4550   5855.0             634.0 2020-12-31 17:42:28   

       requestedDeparture  milesRequested  kWhRequested  WhPerMile  
46086 2021-01-02 06:29:34           130.0         38.87      299.0  
46089 2021-01-02 06:33:29            90.0         28.71      319.0  
46090 2021-01-02 07:23:51            90.0         28.71      319.0  
46091 2021-01-02 16:35:46           250.0         71.50      286.0  
46092 2021-01-02 23:07:37           100.0         27.40      274.0  
...                   ...             ...           ...        ...  
51579 2020-12-30 17:38:39           210.0         59.43      283.0  
51580 2020-12-31 02:17:31           100.0         28.30      283.0  
51582 2020-12-31 06:09:41            50.0         14.50      290.0  
51584 2020-12-31 21:05:43           100.0         28.30      283.0  
51586 2021-01-01 04:16:03           220.0         62.92      286.0  

[1125 rows x 16 

We see that this station is always missing the **doneChargingTime** value.

Now we will take care of these values. We see that we cannot derive this value from other values. Since we have **4087** affected rows out of **65037**, thus **6,3%**, dropping all of those would not result in a significant loss.

In [225]:
df = formatted_df[~formatted_df['doneChargingTime'].isnull()]
df

id            connectionTime  \
0      5e23b149f9af8b5fe4b973cf 2020-01-02 13:08:54+00:00   
1      5e23b149f9af8b5fe4b973d0 2020-01-02 13:36:50+00:00   
2      5e23b149f9af8b5fe4b973d1 2020-01-02 13:56:35+00:00   
3      5e23b149f9af8b5fe4b973d2 2020-01-02 13:59:58+00:00   
4      5e23b149f9af8b5fe4b973d3 2020-01-02 14:00:01+00:00   
...                         ...                       ...   
65032  5d2fbdd3f9af8b4d0dd0d54a 2019-07-01 21:49:12+00:00   
65033  5d2fbdd3f9af8b4d0dd0d54b 2019-07-01 21:58:45+00:00   
65034  5d2fbdd3f9af8b4d0dd0d54c 2019-07-01 22:02:21+00:00   
65035  5d2fbdd3f9af8b4d0dd0d54d 2019-07-01 22:23:44+00:00   
65036  5d2fbdd3f9af8b4d0dd0d54e 2019-07-01 22:36:08+00:00   

                 disconnectTime          doneChargingTime  kWhDelivered  \
0     2020-01-02 19:11:15+00:00 2020-01-02 17:31:35+00:00        25.016   
1     2020-01-02 22:38:21+00:00 2020-01-02 20:18:05+00:00        33.097   
2     2020-01-03 00:39:22+00:00 2020-01-02 16:35:06+00:00         6.521   
3     2020-01-02 16:38:39+00:00 2020-01-02 15:18:45+00:00         2.355   
4     2020-01-02 22:08:40+00:00 2020-01-02 18:17:30+00:00        13.375   
...                         ...                       ...           ...   
65032 2019-07-02 01:37:18+00:00 2019-07-01 23:14:33+00:00         8.399   
65033 2019-07-02 00:39:48+00:00 2019-07-02 00:40:21+00:00        16.864   
65034 2019-07-02 00:58:50+00:00 2019-07-02 00:59:23+00:00        18.335   
65035 2019-07-02 02:03:36+00:00 2019-07-02 02:04:06+00:00        22.815   
65036 2019-07-02 01:10:52+00:00 2019-07-02 00:24:32+00:00        10.241   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
65032  1_1_179_798_2019-07-01 21:49:11.873404       1  AG-3F18  1-1-179-798   
65033  1_1_179_794_2019-07-01 21:58:44.571011       1  AG-3F20  1-1-179-794   
65034  1_1_191_807_2019-07-01 22:02:20.810735       1  AG-4F47  1-1-191-807   
65035  1_1_179_781_2019-07-01 22:23:32.496137       1  AG-3F31  1-1-179-781   
65036  1_1_193_819_2019-07-01 22:35:56.104381       1  AG-1F06  1-1-193-819   

       userID  minutesAvailable          modifiedAt  requestedDeparture  \
0       194.0             463.0 2020-01-02 13:09:39 2020-01-02 20:51:54   
1      4275.0             595.0 2020-01-02 13:37:11 2020-01-02 23:31:50   
2       344.0              60.0 2020-01-02 13:57:17 2020-01-02 14:56:35   
3      1117.0              65.0 2020-01-02 14:00:03 2020-01-02 15:04:58   
4       334.0             504.0 2020-01-02 14:00:13 2020-01-02 22:24:01   
...       ...               ...                 ...                 ...   
65032  1346.0              36.0 2019-07-01 21:49:37 2019-07-01 22:25:12   
65033   364.0             234.0 2019-07-01 21:59:09 2019-07-02 01:52:45   
65034  2050.0             309.0 2019-07-01 22:02:37 2019-07-02 03:11:21   
65035  1626.0              99.0 2019-07-01 22:23:43 2019-07-02 00:02:44   
65036  2299.0             191.0 2019-07-01 22:36:08 2019-07-02 01:47:08   

       milesRequested  kWhRequested  WhPerMile  
0               100.0         25.00      250.0  
1               250.0         70.00      280.0  
2                20.0          8.00      400.0  
3                20.0          8.00      400.0  
4                40.0         16.00      400.0  
...               ...           ...        ...  
65032            80.0         32.00      400.0  
65033           100.0         40.00      400.0  
65034            90.0         29.97      333.0  
65035           170.0         34.00      200.0  
65036 